MatrixSDK/VoIP/CallKit/MXCallKitAdapter.m

```diff
@@ -0,0 +1,285 @@
+    action.contactIdentifier = contactIdentifier;
+    
+    CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
+    [self.callController requestTransaction:transaction completion:^(NSError * _Nullable error) {
+        if (error)
+            return;
+
+        CXCallUpdate *update = [[CXCallUpdate alloc] init];
+        update.remoteHandle = handle;
+        update.localizedCallerName = caller.displayname;
+        update.hasVideo = call.isVideoCall;
+        update.supportsHolding = NO;
+        update.supportsGrouping = NO;
+        update.supportsUngrouping = NO;
+        update.supportsDTMF = NO;
+        
+        [self.provider reportCallWithUUID:callUUID updated:update];
+        
+        [self.calls setObject:call forKey:callUUID];
+    }];
+}
+
+- (void)endCall:(MXCall *)call;
+{
+    MXCallEndReason endReason = call.endReason;
+    
+    if (endReason == MXCallEndReasonHangup)
+    {
+        CXEndCallAction *action = [[CXEndCallAction alloc] initWithCallUUID:call.callUUID];
+        CXTransaction *transaction = [[CXTransaction alloc] initWithAction:action];
+        [self.callController requestTransaction:transaction completion:^(NSError *_Nullable error){}];
+    }
+    else
+    {
+        CXCallEndedReason reason = CXCallEndedReasonFailed;
+        switch (endReason)
+        {
+            case MXCallEndReasonRemoteHangup:
+                reason = CXCallEndedReasonRemoteEnded;
+                break;
+            case MXCallEndReasonHangupElsewhere:
+                reason = CXCallEndedReasonDeclinedElsewhere;
+                break;
+            case MXCallEndReasonBusy:
+            case MXCallEndReasonMissed:
+                reason = CXCallEndedReasonUnanswered;
+                break;
+            default:
+                break;
+        }
+        
+        [self.provider reportCallWithUUID:call.callUUID endedAtDate:nil reason:reason];
+    }
+}
+
+- (void)reportIncomingCall:(MXCall *)call {
+    MXSession *mxSession = call.room.mxSession;
+    MXUser *caller = [mxSession userWithUserId:call.callerId];
+    NSUUID *callUUID = call.callUUID;
+    
+    NSString *handleValue = call.isConferenceCall ? call.room.roomId : call.callerId;
+    CXHandle *handle = [[CXHandle alloc] initWithType:CXHandleTypeGeneric value:handleValue];
+    
+    CXCallUpdate *update = [[CXCallUpdate alloc] init];
+    update.remoteHandle = handle;
+    update.localizedCallerName = caller.displayname;
+    update.hasVideo = call.isVideoCall;
+    update.supportsHolding = NO;
+    update.supportsGrouping = NO;
+    update.supportsUngrouping = NO;
+    update.supportsDTMF = NO;
+    
+    [self.provider reportNewIncomingCallWithUUID:callUUID update:update completion:^(NSError * _Nullable error) {
+        if (error)
+        {
+            [call hangup];
+            return;
+        }
+        
+        self.calls[callUUID] = call;
+        
+#ifdef MX_CALL_STACK_JINGLE
```

Comment:
See my above comment about an abstract MXCallAudioSessionConfigurator.
Instead of this **#ifdef** depending on a VoIP stack, we could use something like ```[callManager.callStack.audioSessionConfigurator configureAudioSessionForVideoCall:]```